datafusion/functions-nested/src/range.rs

```diff
@@ -317,6 +315,14 @@ fn gen_range_date(args: &[ArrayRef], include_upper: bool) -> Result<ArrayRef> {
     let values_builder = Date32Builder::new();
     let mut list_builder = ListBuilder::new(values_builder);
 
+    if start_array.unwrap().is_null(0)
```

Comment:
Actually this function's signature is quite complex, I think we should use `UserDefined` signature to handle all possible types and **coerce** them to the expected one with `coerce_types`. Then we don't need to handle null inside the actual computation here

Reply:
thanks for review. 
replace type coerce with implement coerce_types method.